### PR TITLE
Add downstream test.

### PR DESCRIPTION
**Pull request by CharlesZ-Chen**
As title described. Changes:

- add travis groups: cfi-tests, downstream

- add Ontology as a downstream project of CFI

### CODE REVIEW

File: .travis-build.sh [hunk 1]
```diff
@@ -6,6 +6,40 @@ set -e
...
+if [[ "${GROUP}" == "" ]]; then
+  export GROUP=all
+fi
+
+SLUGOWNER=${TRAVIS_REPO_SLUG%/*}
+if [[ "$SLUGOWNER" == "" ]]; then
+  SLUGOWNER=opprop
+fi
+
+if [[ "${GROUP}" != "cfi-tests" && "${GROUP}" != "downstream" ]]; then
```
Comment (topnessman) [inline .travis-build.sh]: Looks like if GROUP=all case will be considered as bad argument and exit

File: .travis-build.sh [hunk 2]
```diff
@@ -6,6 +6,40 @@ set -e
...
+
+    # Ontology test: 10 minutes
+    echo "Running: (cd .. && git clone --depth 1 https://github.com/opprop/ontology.git)"
+    (cd .. && git clone --depth 1 https://github.com/opprop/ontology.git)
+    echo "... done: (cd .. && git clone --depth 1 https://github.com/opprop/ontology.git)"
+
+    echo "Running: (cd ../ontology && gradle build -x test && ./test-ontology.sh)"
+    (cd ../ontology && gradle build -x test && ./test-ontology.sh)
+    echo "... done: (cd ../ontology && gradle build -x test && ./test-ontology.sh)"
+fi
```
Comment (topnessman) [inline .travis-build.sh]: Add newline would be better

File: scripts/inference [hunk 3]
```diff
@@ -17,9 +17,11 @@ then
     myDir="."
 fi
 
+ROOT=$(cd ${myDir}/../../ && pwd)
```
Comment (topnessman) [inline scripts/inference]: Maybe this is JSR308? I thought ROOT is the CFI project root.

File: src/checkers/inference/InferenceLauncher.java [hunk 4]
```diff
@@ -148,6 +148,7 @@ public void typecheck(String [] javaFiles) {
      */
     public void infer() {
         printStep("Inferring", outStream);
+        printStep("Sys Env: " + System.getenv(), outStream);
```
Comment (topnessman) [inline src/checkers/inference/InferenceLauncher.java]: How many lines does this output? Does it output all the environment variables even irrelevant ones?

File: .travis-build.sh (hunk 1 above)
Comment (CharlesZ-Chen) [reply .travis-build.sh]: Ah, good catch! Will fix this.

File: scripts/inference (hunk 3 above)
Comment (CharlesZ-Chen) [reply scripts/inference]: Sure, will change the name.

File: src/checkers/inference/InferenceLauncher.java (hunk 4 above)
Comment (CharlesZ-Chen) [reply src/checkers/inference/InferenceLauncher.java]: This is just a debug output. I try to see why I cannot pass the travis test.

Will remove this line once I figure out the reason.

**Comment by topnessman** [comment]
Does the downstream test failure fail the upstream CFI build? If yes, it may be not a good idea to add downstream test here.

**Comment by CharlesZ-Chen** [comment]
No, I think the downstream test failed due to we use Docker to run travis test for CFI.

I've been spend some time today to investigate this issue, and somehow I believe it is because our Docker image is Ubuntu 16.04, which cause the failure. (It seems Z3 currently doesn't release a specific version for Ubuntu 16.04, it only has a release for Ubuntu 14.04).

I'll prepare a solution for this, once I figure out a proper docker image for CFI.

**Comment by topnessman** [comment]
This failure is exactly my point of why we shouldn't have downstream test.
We have downstream test only to test make sure that upstream changes in CFI doesn't break the downstream systems, like Ontology and GUT.
But like this example, downstream tests might fail because of lots of reasons other than upstream changes, such as the docker issue here. Or even CF changes might break downstream tests. In these cases, CFI build will fail. This will be a bigger issue if we have more and more downstream build triggered by CFI build. It makes the dependencies very messy.

**Comment by CharlesZ-Chen** [comment]
No, Docker issue is somehow like an environment issue, once it solved, it will not appear again. And thats why I still not merge this PR --- as I still not set up a proper environment for testing downstream projects.

I'm not sure if CF changes might break downstream test, but at least having downstream test is beneficial --- we get earlier notification on changes that break the downstream projects, before merge it into master. It avoid the situation that we merge something in an upstream project, and then suddenly we get a surprise on finding some downstream projects broken.

**Comment by topnessman** [comment]
I understand the initial point of having downstream test. But the cases when downstream tests fail might have unpredictable reasons.(Maybe you just wrote a test that causes a failure to your master. Either Ontology or GUT failure will cause CFI build failure even if it's not due to CFI update). It just makes CFI more brittle in these cases. It's best if we can find a way to trigger the build of downstream but the downstream failure doesn't fail the upstream failure. We can still see the upstream changes in CFI might have issue if downstream tests fail. But the downstream test failure due to other reasons don't cause CFI to fail, which could be unexpected.

**Comment by topnessman** [comment]
I think the ultimate way to improve this is to thoroughly and strongly test CFI instead of adding downstream tests, which may make CFI have lots of false build failures.

**Comment by CharlesZ-Chen** [comment]
> Maybe you just wrote a test that causes a failure to your master. 

Downstream test testing the same content travis do for those downstream projects. I.e. at least we should ensure that downstream projects have a passed master at first.

In your example, you should not merge the failed change into your master at first, as that will not pass the travis test in your project.

**Comment by CharlesZ-Chen** [comment]
Let's discuss this tomorrow in person, Mier.

**Comment by CharlesZ-Chen** [comment]
@wmdietl @topnessman Finally I solved the strange docker issue!

However, I created a new docker image host in my personal docker hub account for making it works.

I still do not know why Michael's docker image does not work, as it seems no essential difference with his docker file and mine (except some extra packages as ontology testing dependences, but those packages does not has the influence on the z3 libpath not found issue).

I've noticed recently docker makes some updates on their environment things, so maybe that might be the reason, as I tried create a new docker image a long time ago, and at that time it didnt work.

Anyway we solve this problem now:)

File: scripts/inference [hunk 5]
```diff
@@ -17,9 +17,11 @@ then
     myDir="."
```
Comment (CharlesZ-Chen) [inline scripts/inference]: change `myDir` to abs path, remove redefinition of `JSR308` in the script.

File: .travis-build.sh [hunk 6]
```diff
@@ -6,6 +6,40 @@ set -e
...
+# Test CF Inference
+if [[ "${GROUP}" == "cfi-tests" || "${GROUP}" == "all" ]]; then
+    ant -f tests.xml run-tests
+fi
+
+# Downstream tests
+if [[ "${GROUP}" == "downstream" || "${GROUP}" == "all" ]]; then
+
+    # Ontology test: 10 minutes
+    echo "Running: (cd .. && git clone --depth 1 https://github.com/opprop/ontology.git)"
```
Comment (CharlesZ-Chen) [inline .travis-build.sh]: change `opprop` to `SLUGOWNER`

File: .travis-build.sh [hunk 7]
```diff
@@ -6,6 +6,40 @@ set -e
...
 . ./.travis-build-without-test.sh
 
-ant -f tests.xml run-tests
+# Test CF Inference
+if [[ "${GROUP}" == "cfi-tests" || "${GROUP}" == "all" ]]; then
+    ant -f tests.xml run-tests
+fi
+
+# Downstream tests
+if [[ "${GROUP}" == "downstream" || "${GROUP}" == "all" ]]; then
```
Comment (CharlesZ-Chen) [inline .travis-build.sh]: Do condition check on `SLUGOWNER`, only perform down stream test on ontology in opprop.

**Pull request merged**